sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousProgressReporter.scala

```diff
@@ -0,0 +1,182 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.streaming.continuous
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+import org.apache.spark.rpc.RpcEndpointRef
+import org.apache.spark.sql.catalyst.util.DateTimeUtils._
+import org.apache.spark.sql.execution.streaming.{ ProgressReporter, StreamProgress}
+import org.apache.spark.sql.sources.v2.reader.streaming.SparkDataStream
+import org.apache.spark.sql.streaming.{SinkProgress, SourceProgress, StreamingQueryProgress}
+
+trait ContinuousProgressReporter extends ProgressReporter {
+
+  protected def epochEndpoint: RpcEndpointRef
+
+  private var earliestEpochId: Long = -1
+
+  private val currentDurationsMs =
+    new mutable.HashMap[Long, (Long, mutable.HashMap[String, Long])]()
+  private val recordDurationMs = new mutable.HashMap[String, Long]()
+
+  private val currentTriggerStartOffsets: mutable.HashMap[Long, Map[SparkDataStream, String]] =
+    new mutable.HashMap[Long, Map[SparkDataStream, String]]()
+  private val currentTriggerEndOffsets: mutable.HashMap[Long, Map[SparkDataStream, String]] =
+    new mutable.HashMap[Long, Map[SparkDataStream, String]]()
+
+  // TODO: Restore this from the checkpoint when possible.
+  private var lastTriggerStartTimestamp = -1L
+
+  // Local timestamps and counters.
+  private var currentTriggerStartTimestamp = -1L
+
+  private val noDataProgressEventInterval =
+    sparkSession.sessionState.conf.streamingNoDataProgressEventInterval
+
+  // The timestamp we report an event that has no input data
+  private var lastNoDataProgressEventTime = Long.MinValue
+
+  /** Begins recording statistics about query progress for a given trigger. */
+  override protected def startTrigger(): Unit = {
+    logDebug("Starting Trigger Calculation")
+    if (earliestEpochId == -1) {
+      earliestEpochId = currentBatchId
+    }
+    checkQueueBoundaries()
+    lastTriggerStartTimestamp = currentTriggerStartTimestamp
+    currentTriggerStartTimestamp = triggerClock.getTimeMillis()
+    currentDurationsMs.put(currentBatchId,
+      (currentTriggerStartTimestamp, new mutable.HashMap[String, Long]()))
+  }
+
+  /** Finalizes the query progress and adds it to list of recent status updates. */
+  protected def finishTrigger(hasNewData: Boolean, epochId: Long, epochStats: EpochStats): Unit = {
+    assert(currentTriggerStartOffsets.get(epochId).isDefined
+      && currentTriggerEndOffsets.get(epochId).isDefined
+      && currentDurationsMs.get(epochId).isDefined)
+    val currentTriggerStartTimestamp = currentDurationsMs(epochId)._1
+    val currentTriggerEndTimestamp = triggerClock.getTimeMillis()
+
+    val executionStats = extractExecutionStats(hasNewData, Some(epochStats))
```

Comment:
Since the origin `extractExecutionStats` method collects information from the SQL metrics (e.g. state operator metrics and watermark) which is reported only when task ended for now,  is there a plan to handle this part of data?